### ISSUE TITLE
Package upgrade fails because Flutter uses non-standard version naming for hotfixes

### ISSUE THREAD

**Issue by apaatsio** [flutter/flutter#32044]
## Steps to Reproduce

1. Take any existing project or create a new one / `flutter create foo`
2. Edit `pubspec.yaml` in the project so that it has the section
```yaml
environment:
  sdk: ">=2.1.0 <3.0.0"
  flutter: ">=1.5.4 <2.0.0"
```
3. Switch to 1.5.4-hotfix.2 in the project directory / `flutter version 1.5.4-hotfix.2` (This command will also run `flutter packages upgrade`)

## Expected result

Packages are upgraded successfully.

## Actual result

Package upgrade fails.

## Log
```
$ flutter version 1.5.4-hotfix.2
Building flutter tool...
<!-- SNIP -->
Flutter 1.5.4-hotfix.2 • channel unknown • unknown source
Framework • revision 7a4c33425d (4 days ago) • 2019-04-29 11:05:24 -0700
Engine • revision 52c7a1e849
Tools • Dart 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)

The current Flutter SDK version is 1.5.4-hotfix.2.                      
                                                                        
Because foo requires Flutter SDK version >=1.5.4 <2.0.0, version solving failed.
Running "flutter packages upgrade" in foo...                            
pub upgrade failed (1)
```

## Root cause

The upgrade command thinks that `1.5.4-hotfix.2` is older than `1.5.4` because Flutter uses non-standard version naming. [Hotfixes should use + as a separator and pre-releases -](https://www.dartlang.org/tools/pub/pubspec#version) but Flutter uses - for both.

That is, `1.5.4-hotfix.2` should be named `1.5.4+hotfix.2` for `flutter packages upgrade`/`pub upgrade` to work as expected.

**Comment by Hixie** [comment]
I believe this is fixed now on master. cc @jonahwilliams for confirmation.

**Comment by BondarenkoStas** [comment]
already in stable (merged to 1.6.0)
